File: stock-portfolio-advisor-backend/src/main/java/com/bryanphone/stockadvisor/utils/CompressionUtil.java
package com.bryanphone.stockadvisor.utils;

import lombok.extern.java.Log;

import java.io.ByteArrayOutputStream;
import java.util.zip.Deflater;
import java.util.zip.Inflater;

@Log
public class CompressionUtil {
    // Compress a string into a byte array using gzip
    public static byte[] compress(String data) {
        if (data == null || data.isEmpty()) {
            return new byte[0];
        }

        Deflater deflater = new Deflater();
        deflater.setInput(data.getBytes());
        deflater.finish();

        ByteArrayOutputStream outputStream = new ByteArrayOutputStream(data.length());
        byte[] buffer = new byte[1024];
        while (!deflater.finished()) {
            int count = deflater.deflate(buffer);
            outputStream.write(buffer, 0, count);
        }
        deflater.end();

        return outputStream.toByteArray();
    }

    // Decompress a byte array into a string using gzip
    public static String decompress(byte[] compressedData) {
        if (compressedData == null || compressedData.length == 0) {
            return "";
        }

        Inflater inflater = new Inflater();
        inflater.setInput(compressedData);

        ByteArrayOutputStream outputStream = new ByteArrayOutputStream(compressedData.length);
        byte[] buffer = new byte[1024];
        try {
            while (!inflater.finished()) {
                int count = inflater.inflate(buffer);
                outputStream.write(buffer, 0, count);
            }
        } catch (Exception e) {
            throw new RuntimeException("Failed to decompress data", e);
        } finally {
            inflater.end();
        }

        return outputStream.toString();
    }
}

File: stock-portfolio-advisor-bk-nodejs/src/utils/CompressionUtil.js
/**
 * Utility class for compression and decompression of data
 * Converted from Java implementation to Node.js
 */
const zlib = require('zlib');
const util = require('util');

// Promisify zlib methods for async/await usage
// Using gzip instead of deflate for better PostgreSQL compatibility
const gzip = util.promisify(zlib.gzip);
const gunzip = util.promisify(zlib.gunzip);

class CompressionUtil {
  /**
   * Compress a string into a Buffer using zlib gzip
   * @param {string} data - String data to compress
   * @returns {Promise<Buffer>} - Compressed data as Buffer
   */
  static async compress(data) {
    if (data == null || data === '') {
      return Buffer.alloc(0);
    }
    
    try {
      // Use gzip instead of deflate for better PostgreSQL compatibility
      const compressed = await gzip(Buffer.from(data));
      return compressed;
    } catch (error) {
      console.error('Compression error:', error);
      throw new Error('Failed to compress data: ' + error.message);
    }
  }

  /**
   * Decompress a Buffer into a string using zlib gunzip
   * @param {Buffer} compressedData - Compressed data as Buffer
   * @returns {Promise<string>} - Decompressed string
   */
  static async decompress(compressedData) {
    if (!compressedData || !Buffer.isBuffer(compressedData) || compressedData.length === 0) {
      return '';
    }
    
    try {
      // Use gunzip instead of inflate for better PostgreSQL compatibility
      const result = await gunzip(compressedData);
      return result.toString();
    } catch (error) {
      console.error('Decompression error:', error);
      throw new Error('Failed to decompress data: ' + error.message);
    }
  }

  /**
   * Safely decompress data that might not be compressed
   * @param {Buffer|any} data - Data to decompress
   * @returns {Promise<string>} - Decompressed string or original data as string
   */
  static async safeDecompress(data) {
    if (!data) {
      return '';
    }
    
    // Ensure we have a buffer
    let buffer;
    if (!Buffer.isBuffer(data)) {
      // Handle numeric values or other types by converting to string first
      const stringData = typeof data === 'number' ? String(data) : data.toString();
      buffer = Buffer.from(stringData);
    } else {
      buffer = data;
    }
    
    try {
      // Try to decompress using gunzip
      const decompressed = await gunzip(buffer);
      return decompressed.toString();
    } catch (error) {
      // Try to detect if this is a gzip error or just uncompressed data
      if (error.code === 'Z_DATA_ERROR') {
        // This is likely uncompressed data
        console.log('Data is not compressed, using as-is');
        return buffer.toString();
      }
      
      // For other errors, log and still try to return something useful
      console.error('Error in safeDecompress:', error.message);
      return buffer.toString();
    }
  }
}

module.exports = CompressionUtil;
